archive/backup.go

```diff
@@ -171,59 +178,104 @@ func writeMetadata(writer io.Writer, logger hclog.Logger, to uint64, toHash type
 	return err
 }
 
-func processExportStream(
-	stream proto.System_ExportClient,
+func processExport(
+	ctx context.Context,
+	clt proto.SystemClient,
 	logger hclog.Logger,
 	writer io.Writer,
 	targetFrom, targetTo uint64,
 ) (uint64, uint64, error) {
-	var from, to, total uint64 = 0, 0, 0
+	var from, to, current, total uint64 = targetFrom, targetTo, targetFrom, 0
 
-	showProgress := func(event *proto.ExportEvent) {
-		num := event.To - event.From
-		total += num
-		expectedTo := targetTo
-
-		if targetTo == 0 {
-			expectedTo = event.Latest
-		}
+	if (targetTo - targetFrom + 1) == 0 {
+		return 0, 0, ErrBlockRange
+	}
 
-		expectedTotal := expectedTo - targetFrom
-		progress := 100 * (float64(event.To) - float64(targetFrom)) / float64(expectedTotal)
+	showProgress := func(block *types.Block) {
+		current = block.Header.Number
+		total += 1
 
+		progress := 100 * (float64(total) / float64(targetTo-targetFrom+1))
 		logger.Info(
-			fmt.Sprintf("%d blocks are written", num),
-			"total", total,
-			"from", targetFrom,
-			"to", expectedTo,
+			fmt.Sprintf("%d blocks are written", total),
+			"from", from,
+			"to", to,
+			"height", current,
 			"progress", fmt.Sprintf("%.2f%%", progress),
 		)
 	}
 
-	firstBlok := true
+	maxFetchBlockNum := runtime.NumCPU() * 2
 
-	for {
-		event, err := stream.Recv()
-		if errors.Is(io.EOF, err) || status.Code(err) == codes.Canceled {
-			return from, to, nil
-		}
+	// fetch blocks from channel
+	// block data channel: [from, to]
+	blockCh := make(chan *types.Block, maxFetchBlockNum)
 
-		if err != nil {
-			return from, to, err
-		}
+	go func(from uint64, to uint64) {
+		defer func() {
+			blockCh <- nil
+		}()
 
-		// tips: writer.Write() not necessarily write all data, use io.Copy() instead
-		if _, err := io.Copy(writer, bytes.NewBuffer(event.Data)); err != nil {
-			return from, to, err
-		}
+		for num := from; num <= to; num++ {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+
+				resp, err := clt.BlockByNumber(context.Background(), &proto.BlockByNumberRequest{Number: num})
+				if status.Code(err) == codes.Canceled {
+					return
+				}
+
+				if err != nil {
+					logger.Error("Failed to fetch block", "Number", num, "err", err)
+					time.Sleep(1 * time.Second)
+
+					continue
```

Comment:
Why not return?
You should only retry if gRPC returns `ResourceExhausted`. Other failures are not recoverable.

Reply:
Now the backup method cannot be continued from the middle, if the grpc endpoint is after load balancing (e.g. envoy), then any failure can be retried instead of quitting halfway through the backup with an error.